File: frontend-mobile/mobile/packages/divine_camera/android/src/main/kotlin/co/openvine/divine_camera/DivineCameraPlugin.kt
// ABOUTME: Main Flutter plugin entry point for Android camera operations
// ABOUTME: Handles method channel communication and delegates to CameraController

package co.openvine.divine_camera

import android.app.Activity
import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.view.TextureRegistry

/** DivineCameraPlugin */
class DivineCameraPlugin :
    FlutterPlugin,
    MethodCallHandler,
    ActivityAware {

    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private lateinit var textureRegistry: TextureRegistry
    private var activity: Activity? = null
    private var cameraController: CameraController? = null
    private var volumeKeyHandler: VolumeKeyHandler? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "divine_camera")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
        textureRegistry = flutterPluginBinding.textureRegistry
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }

            "initializeCamera" -> {
                val lens = call.argument<String>("lens") ?: "back"
                val videoQuality = call.argument<String>("videoQuality") ?: "fhd"
                val enableScreenFlash = call.argument<Boolean>("enableScreenFlash") ?: true
                val mirrorFrontCameraOutput = call.argument<Boolean>("mirrorFrontCameraOutput") ?: true
                initializeCamera(lens, videoQuality, enableScreenFlash, mirrorFrontCameraOutput, result)
            }

            "disposeCamera" -> {
                disposeCamera(result)
            }

            "setFlashMode" -> {
                val mode = call.argument<String>("mode") ?: "off"
                setFlashMode(mode, result)
            }

            "setFocusPoint" -> {
                val x = call.argument<Double>("x") ?: 0.5
                val y = call.argument<Double>("y") ?: 0.5
                setFocusPoint(x.toFloat(), y.toFloat(), result)
            }

            "setExposurePoint" -> {
                val x = call.argument<Double>("x") ?: 0.5
                val y = call.argument<Double>("y") ?: 0.5
                setExposurePoint(x.toFloat(), y.toFloat(), result)
            }

            "cancelFocusAndMetering" -> {
                cancelFocusAndMetering(result)
            }

            "setZoomLevel" -> {
                val level = call.argument<Double>("level") ?: 1.0
                setZoomLevel(level.toFloat(), result)
            }

            "switchCamera" -> {
                val lens = call.argument<String>("lens") ?: "back"
                switchCamera(lens, result)
            }

            "startRecording" -> {
                val maxDurationMs = call.argument<Int>("maxDurationMs")
                val useCache = call.argument<Boolean>("useCache") ?: true
                val outputDirectory = call.argument<String>("outputDirectory")
                startRecording(maxDurationMs, useCache, outputDirectory, result)
            }

            "stopRecording" -> {
                stopRecording(result)
            }

            "pausePreview" -> {
                pausePreview(result)
            }

            "resumePreview" -> {
                resumePreview(result)
            }

            "getCameraState" -> {
                getCameraState(result)
            }

            "setRemoteRecordControlEnabled" -> {
                val enabled = call.argument<Boolean>("enabled") ?: false
                setRemoteRecordControlEnabled(enabled, result)
            }

            "setVolumeKeysEnabled" -> {
                val enabled = call.argument<Boolean>("enabled") ?: true
                setVolumeKeysEnabled(enabled, result)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    private fun initializeCamera(lens: String, videoQuality: String, enableScreenFlash: Boolean, mirrorFrontCameraOutput: Boolean, result: Result) {
        val currentActivity = activity
        if (currentActivity == null) {
            result.error("NO_ACTIVITY", "Activity not available", null)
            return
        }

        try {
            cameraController?.release()
            cameraController = CameraController(
                context = context,
                activity = currentActivity,
                textureRegistry = textureRegistry
            )

            // Set up auto-stop listener to notify Flutter
            cameraController?.onAutoStopListener = { recordingResult ->
                channel.invokeMethod("onRecordingAutoStopped", recordingResult)
            }

            cameraController?.initialize(lens, videoQuality, enableScreenFlash, mirrorFrontCameraOutput) { state, error ->
                if (error != null) {
                    result.error("INIT_ERROR", error, null)
                } else {
                    result.success(state)
                }
            }
        } catch (e: Exception) {
            result.error("INIT_EXCEPTION", e.message, e.stackTraceToString())
        }
    }

    private fun disposeCamera(result: Result) {
        try {
            volumeKeyHandler?.release()
            volumeKeyHandler = null
            cameraController?.release()
            cameraController = null
            result.success(null)
        } catch (e: Exception) {
            result.error("DISPOSE_ERROR", e.message, null)
        }
    }

    private fun setFlashMode(mode: String, result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            val success = controller.setFlashMode(mode)
            result.success(success)
        } catch (e: Exception) {
            result.error("FLASH_ERROR", e.message, null)
        }
    }

    private fun setFocusPoint(x: Float, y: Float, result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            val success = controller.setFocusPoint(x, y)
            result.success(success)
        } catch (e: Exception) {
            result.error("FOCUS_ERROR", e.message, null)
        }
    }

    private fun setExposurePoint(x: Float, y: Float, result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            val success = controller.setExposurePoint(x, y)
            result.success(success)
        } catch (e: Exception) {
            result.error("EXPOSURE_ERROR", e.message, null)
        }
    }

    private fun cancelFocusAndMetering(result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            val success = controller.cancelFocusAndMetering()
            result.success(success)
        } catch (e: Exception) {
            result.error("FOCUS_ERROR", e.message, null)
        }
    }

    private fun setZoomLevel(level: Float, result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            val success = controller.setZoomLevel(level)
            result.success(success)
        } catch (e: Exception) {
            result.error("ZOOM_ERROR", e.message, null)
        }
    }

    private fun switchCamera(lens: String, result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            // Suppress volume/Bluetooth triggers during camera switch.
            // Camera reconfiguration can cause connected Bluetooth devices
            // to send spurious play/pause events that would start recording.
            volumeKeyHandler?.suppressTemporarily(3000)

            controller.switchCamera(lens) { state, error ->
                if (error != null) {
                    result.error("SWITCH_ERROR", error, null)
                } else {
                    result.success(state)
                }
            }
        } catch (e: Exception) {
            result.error("SWITCH_EXCEPTION", e.message, null)
        }
    }

    private fun startRecording(maxDurationMs: Int?, useCache: Boolean, outputDirectory: String?, result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            controller.startRecording(maxDurationMs, useCache, outputDirectory) { error ->
                if (error != null) {
                    result.error("RECORD_START_ERROR", error, null)
                } else {
                    result.success(null)
                }
            }
        } catch (e: Exception) {
            result.error("RECORD_START_EXCEPTION", e.message, null)
        }
    }

    private fun stopRecording(result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            controller.stopRecording { recordingResult, error ->
                if (error != null) {
                    result.error("RECORD_STOP_ERROR", error, null)
                } else {
                    result.success(recordingResult)
                }
            }
        } catch (e: Exception) {
            result.error("RECORD_STOP_EXCEPTION", e.message, null)
        }
    }

    private fun pausePreview(result: Result) {
        try {
            cameraController?.pausePreview()
            result.success(null)
        } catch (e: Exception) {
            result.error("PAUSE_ERROR", e.message, null)
        }
    }

    private fun resumePreview(result: Result) {
        try {
            cameraController?.resumePreview { state, error ->
                if (error != null) {
                    result.error("RESUME_ERROR", error, null)
                } else {
                    result.success(state)
                }
            }
        } catch (e: Exception) {
            result.error("RESUME_EXCEPTION", e.message, null)
        }
    }

    private fun getCameraState(result: Result) {
        val controller = cameraController
        if (controller == null) {
            result.error("NOT_INITIALIZED", "Camera not initialized", null)
            return
        }
        try {
            result.success(controller.getCameraState())
        } catch (e: Exception) {
            result.error("STATE_ERROR", e.message, null)
        }
    }

    private fun setRemoteRecordControlEnabled(enabled: Boolean, result: Result) {
        try {
            if (enabled) {
                if (volumeKeyHandler == null) {
                    volumeKeyHandler = VolumeKeyHandler(context, activity) { triggerType ->
                        // Send trigger event to Flutter on main thread
                        android.os.Handler(android.os.Looper.getMainLooper()).post {
                            channel.invokeMethod("onRemoteRecordTrigger", triggerType)
                        }
                    }
                }
                val success = volumeKeyHandler?.enable() ?: false
                result.success(success)
            } else {
                volumeKeyHandler?.disable()
                result.success(true)
            }
        } catch (e: Exception) {
            result.error("REMOTE_CONTROL_ERROR", e.message, null)
        }
    }

    private fun setVolumeKeysEnabled(enabled: Boolean, result: Result) {
        try {
            volumeKeyHandler?.setVolumeKeysEnabled(enabled)
            result.success(true)
        } catch (e: Exception) {
            result.error("VOLUME_KEYS_ERROR", e.message, null)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        volumeKeyHandler?.release()
        volumeKeyHandler = null
        cameraController?.release()
        cameraController = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}

File: frontend-mobile/mobile/packages/divine_camera/ios/Classes/DivineCameraPlugin.swift
// ABOUTME: Main Flutter plugin entry point for iOS camera operations
// ABOUTME: Handles method channel communication and delegates to CameraController

import Flutter
import UIKit

public class DivineCameraPlugin: NSObject, FlutterPlugin {
    private var cameraController: CameraController?
    private var textureRegistry: FlutterTextureRegistry?
    private var messenger: FlutterBinaryMessenger?
    private var methodChannel: FlutterMethodChannel?
    private var volumeKeyHandler: VolumeKeyHandler?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "divine_camera", binaryMessenger: registrar.messenger())
        let instance = DivineCameraPlugin()
        instance.textureRegistry = registrar.textures()
        instance.messenger = registrar.messenger()
        instance.methodChannel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        // Listen for auto-stop events from CameraController
        NotificationCenter.default.addObserver(
            instance,
            selector: #selector(instance.handleAutoStop(_:)),
            name: NSNotification.Name("DivineCameraAutoStop"),
            object: nil
        )
    }
    
    @objc private func handleAutoStop(_ notification: Notification) {
        guard let userInfo = notification.userInfo as? [String: Any] else { return }
        
        // Invoke method channel to notify Flutter of auto-stop
        DispatchQueue.main.async { [weak self] in
            self?.methodChannel?.invokeMethod("onRecordingAutoStopped", arguments: userInfo)
        }
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
            
        case "initializeCamera":
            let args = call.arguments as? [String: Any] ?? [:]
            let lens = args["lens"] as? String ?? "back"
            let videoQuality = args["videoQuality"] as? String ?? "fhd"
            let enableScreenFlash = args["enableScreenFlash"] as? Bool ?? true
            let mirrorFrontCameraOutput = args["mirrorFrontCameraOutput"] as? Bool ?? true
            initializeCamera(lens: lens, videoQuality: videoQuality, enableScreenFlash: enableScreenFlash, mirrorFrontCameraOutput: mirrorFrontCameraOutput, result: result)
            
        case "disposeCamera":
            disposeCamera(result: result)
            
        case "setFlashMode":
            let args = call.arguments as? [String: Any] ?? [:]
            let mode = args["mode"] as? String ?? "off"
            setFlashMode(mode: mode, result: result)
            
        case "setFocusPoint":
            let args = call.arguments as? [String: Any] ?? [:]
            let x = args["x"] as? Double ?? 0.5
            let y = args["y"] as? Double ?? 0.5
            setFocusPoint(x: x, y: y, result: result)
            
        case "setExposurePoint":
            let args = call.arguments as? [String: Any] ?? [:]
            let x = args["x"] as? Double ?? 0.5
            let y = args["y"] as? Double ?? 0.5
            setExposurePoint(x: x, y: y, result: result)
            
        case "cancelFocusAndMetering":
            cancelFocusAndMetering(result: result)
            
        case "setZoomLevel":
            let args = call.arguments as? [String: Any] ?? [:]
            let level = args["level"] as? Double ?? 1.0
            setZoomLevel(level: level, result: result)
            
        case "switchCamera":
            let args = call.arguments as? [String: Any] ?? [:]
            let lens = args["lens"] as? String ?? "back"
            switchCamera(lens: lens, result: result)
            
        case "startRecording":
            let args = call.arguments as? [String: Any] ?? [:]
            let maxDurationMs = args["maxDurationMs"] as? Int
            let useCache = args["useCache"] as? Bool ?? true
            let outputDirectory = args["outputDirectory"] as? String
            startRecording(maxDurationMs: maxDurationMs, useCache: useCache, outputDirectory: outputDirectory, result: result)
            
        case "stopRecording":
            stopRecording(result: result)
            
        case "pausePreview":
            pausePreview(result: result)
            
        case "resumePreview":
            resumePreview(result: result)
            
        case "getCameraState":
            getCameraState(result: result)
            
        case "setRemoteRecordControlEnabled":
            let args = call.arguments as? [String: Any] ?? [:]
            let enabled = args["enabled"] as? Bool ?? false
            setRemoteRecordControlEnabled(enabled: enabled, result: result)
            
        case "setVolumeKeysEnabled":
            let args = call.arguments as? [String: Any] ?? [:]
            let enabled = args["enabled"] as? Bool ?? true
            setVolumeKeysEnabled(enabled: enabled, result: result)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func initializeCamera(lens: String, videoQuality: String, enableScreenFlash: Bool, mirrorFrontCameraOutput: Bool, result: @escaping FlutterResult) {
        guard let registry = textureRegistry else {
            result(FlutterError(code: "NO_REGISTRY", message: "Texture registry not available", details: nil))
            return
        }
        
        cameraController?.release()
        cameraController = CameraController(textureRegistry: registry)
        
        cameraController?.initialize(lens: lens, videoQuality: videoQuality, enableScreenFlash: enableScreenFlash, mirrorFrontCameraOutput: mirrorFrontCameraOutput) { [weak self] state, error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "INIT_ERROR", message: error, details: nil))
                } else {
                    result(state)
                }
            }
        }
    }
    
    private func disposeCamera(result: @escaping FlutterResult) {
        volumeKeyHandler?.release()
        volumeKeyHandler = nil
        cameraController?.release()
        cameraController = nil
        result(nil)
    }
    
    private func setRemoteRecordControlEnabled(enabled: Bool, result: @escaping FlutterResult) {
        if enabled {
            if volumeKeyHandler == nil {
                volumeKeyHandler = VolumeKeyHandler { [weak self] triggerType in
                    // Send trigger event to Flutter on main thread
                    DispatchQueue.main.async {
                        self?.methodChannel?.invokeMethod("onRemoteRecordTrigger", arguments: triggerType)
                    }
                }
            }
            let success = volumeKeyHandler?.enable() ?? false
            result(success)
        } else {
            volumeKeyHandler?.disable()
            result(true)
        }
    }
    
    private func setVolumeKeysEnabled(enabled: Bool, result: @escaping FlutterResult) {
        volumeKeyHandler?.setVolumeKeysEnabled(enabled)
        result(true)
    }
    
    private func setFlashMode(mode: String, result: @escaping FlutterResult) {
        guard let controller = cameraController else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        let success = controller.setFlashMode(mode: mode)
        result(success)
    }
    
    private func setFocusPoint(x: Double, y: Double, result: @escaping FlutterResult) {
        guard let controller = cameraController else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        let success = controller.setFocusPoint(x: CGFloat(x), y: CGFloat(y))
        result(success)
    }
    
    private func setExposurePoint(x: Double, y: Double, result: @escaping FlutterResult) {
        guard let controller = cameraController else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        let success = controller.setExposurePoint(x: CGFloat(x), y: CGFloat(y))
        result(success)
    }
    
    private func cancelFocusAndMetering(result: @escaping FlutterResult) {
        guard let controller = cameraController else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        let success = controller.cancelFocusAndMetering()
        result(success)
    }
    
    private func setZoomLevel(level: Double, result: @escaping FlutterResult) {
        guard let controller = cameraController else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        let success = controller.setZoomLevel(level: CGFloat(level))
        result(success)
    }
    
    private func switchCamera(lens: String, result: @escaping FlutterResult) {
        guard let controller = cameraController else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        
        // Suppress Bluetooth triggers during camera switch.
        // iOS re-evaluates audio routing on AVCaptureSession reconfiguration,
        // which can cause connected Bluetooth devices (Apple Watch, AirPods)
        // to send spurious play/pause events that would restart recording.
        volumeKeyHandler?.suppressTemporarily(forSeconds: 3.0)
        
        controller.switchCamera(lens: lens) { state, error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "SWITCH_ERROR", message: error, details: nil))
                } else {
                    result(state)
                }
            }
        }
    }
    
    private func startRecording(maxDurationMs: Int?, useCache: Bool, outputDirectory: String?, result: @escaping FlutterResult) {
        guard let controller = cameraController else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        
        controller.startRecording(maxDurationMs: maxDurationMs, useCache: useCache, outputDirectory: outputDirectory) { error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "RECORD_START_ERROR", message: error, details: nil))
                } else {
                    result(nil)
                }
            }
        }
    }
    
    private func stopRecording(result: @escaping FlutterResult) {
        guard let controller = cameraController else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        
        controller.stopRecording { recordingResult, error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "RECORD_STOP_ERROR", message: error, details: nil))
                } else {
                    result(recordingResult)
                }
            }
        }
    }
    
    private func pausePreview(result: @escaping FlutterResult) {
        cameraController?.pausePreview()
        result(nil)
    }
    
    private func resumePreview(result: @escaping FlutterResult) {
        cameraController?.resumePreview { state, error in
            DispatchQueue.main.async {
                if let error = error {
                    result(FlutterError(code: "RESUME_ERROR", message: error, details: nil))
                } else {
                    result(state)
                }
            }
        }
    }
    
    private func getCameraState(result: @escaping FlutterResult) {
        guard let controller = cameraController else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "Camera not initialized", details: nil))
            return
        }
        result(controller.getCameraState())
    }
}
